net/dns: OSConfigurator.SetDNS with systemd-resolved errors Unknown method 'SetLinkDefaultRoute'

### What is the issue?

In `net/dns.OSConfigurator` method `SetDNS` throws an error `setLinkDefaultRoute: Unknown method 'SetLinkDefaultRoute' or interface 'org.freedesktop.resolve1.Manager'.`

I am not a tailscale user, thus I am not sure how this affects regular users.


## System info

Kubuntu 18.04.6

dns: [rc=resolved nm=yes nm-resolved=yes nm-safe=no ret=systemd-resolved]

```
$ systemd-resolve --version
systemd 237
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN -PCRE2 default-hierarchy=hybrid
```

```
$ NetworkManager -V
1.10.14
```

```
$ gdbus introspect --system \
  --dest org.freedesktop.resolve1 \
  --object-path /org/freedesktop/resolve1
```

<details>
 <summary>Output of introspect</summary>

```
node /org/freedesktop/resolve1 {
  interface org.freedesktop.DBus.Peer {
    methods:
      Ping();
      GetMachineId(out s machine_uuid);
    signals:
    properties:
  };
  interface org.freedesktop.DBus.Introspectable {
    methods:
      Introspect(out s data);
    signals:
    properties:
  };
  interface org.freedesktop.DBus.Properties {
    methods:
      Get(in  s interface,
          in  s property,
          out v value);
      GetAll(in  s interface,
             out a{sv} properties);
      Set(in  s interface,
          in  s property,
          in  v value);
    signals:
      PropertiesChanged(s interface,
                        a{sv} changed_properties,
                        as invalidated_properties);
    properties:
  };
  interface org.freedesktop.resolve1.Manager {
    methods:
      ResolveHostname(in  i arg_0,
                      in  s arg_1,
                      in  i arg_2,
                      in  t arg_3,
                      out a(iiay) arg_4,
                      out s arg_5,
                      out t arg_6);
      ResolveAddress(in  i arg_0,
                     in  i arg_1,
                     in  ay arg_2,
                     in  t arg_3,
                     out a(is) arg_4,
                     out t arg_5);
      ResolveRecord(in  i arg_0,
                    in  s arg_1,
                    in  q arg_2,
                    in  q arg_3,
                    in  t arg_4,
                    out a(iqqay) arg_5,
                    out t arg_6);
      ResolveService(in  i arg_0,
                     in  s arg_1,
                     in  s arg_2,
                     in  s arg_3,
                     in  i arg_4,
                     in  t arg_5,
                     out a(qqqsa(iiay)s) arg_6,
                     out aay arg_7,
                     out s arg_8,
                     out s arg_9,
                     out s arg_10,
                     out t arg_11);
      @org.freedesktop.systemd1.Privileged("true")
      ResetStatistics();
      @org.freedesktop.systemd1.Privileged("true")
      FlushCaches();
      @org.freedesktop.systemd1.Privileged("true")
      ResetServerFeatures();
      GetLink(in  i arg_0,
              out o arg_1);
      @org.freedesktop.systemd1.Privileged("true")
      SetLinkDNS(in  i arg_0,
                 in  a(iay) arg_1);
      @org.freedesktop.systemd1.Privileged("true")
      SetLinkDomains(in  i arg_0,
                     in  a(sb) arg_1);
      @org.freedesktop.systemd1.Privileged("true")
      SetLinkLLMNR(in  i arg_0,
                   in  s arg_1);
      @org.freedesktop.systemd1.Privileged("true")
      SetLinkMulticastDNS(in  i arg_0,
                          in  s arg_1);
      @org.freedesktop.systemd1.Privileged("true")
      SetLinkDNSSEC(in  i arg_0,
                    in  s arg_1);
      @org.freedesktop.systemd1.Privileged("true")
      SetLinkDNSSECNegativeTrustAnchors(in  i arg_0,
                                        in  as arg_1);
      @org.freedesktop.systemd1.Privileged("true")
      RevertLink(in  i arg_0);
      RegisterService(in  s arg_0,
                      in  s arg_1,
                      in  s arg_2,
                      in  q arg_3,
                      in  q arg_4,
                      in  q arg_5,
                      in  aa{say} arg_6,
                      out o arg_7);
      UnregisterService(in  o arg_0);
    signals:
    properties:
      @org.freedesktop.DBus.Property.EmitsChangedSignal("false")
      readonly s LLMNRHostname = 'MyBook';
      @org.freedesktop.DBus.Property.EmitsChangedSignal("false")
      readonly a(iiay) DNS = [(13, 2, [0x7f, 0x00, 0x00, 0x42]), (4, 2, [0xc0, 0xa8, 0x01, 0x01])];
      @org.freedesktop.DBus.Property.EmitsChangedSignal("false")
      readonly a(isb) Domains = [(13, 'mydomain', true), (4, '.', true)];
      @org.freedesktop.DBus.Property.EmitsChangedSignal("false")
      readonly (tt) TransactionStatistics = (2, 1868);
      @org.freedesktop.DBus.Property.EmitsChangedSignal("false")
      readonly (ttt) CacheStatistics = (119, 945, 948);
      @org.freedesktop.DBus.Property.EmitsChangedSignal("false")
      readonly (tttt) DNSSECStatistics = (0, 0, 0, 0);
      @org.freedesktop.DBus.Property.EmitsChangedSignal("false")
      readonly b DNSSECSupported = false;
      @org.freedesktop.DBus.Property.EmitsChangedSignal("false")
      readonly as DNSSECNegativeTrustAnchors = [.....];
      @org.freedesktop.DBus.Property.EmitsChangedSignal("false")
      readonly s DNSStubListener = 'yes';
  };
  node dnssd {
  };
  node link {
  };
};
```

</details>


### Steps to reproduce

Easiest way for me was to write "go test", but probably this error should appear in regular tailscale logs.


Content of `resolved_test.go` in `net/dns` directory
<details>

```go
package dns

import (
	"fmt"
	"log"
	"testing"
	"time"

	"inet.af/netaddr"
	"tailscale.com/util/dnsname"
)

func TestNewManager(t *testing.T) {
	const (
		interfaceName = "tt0"
		domainZone    = "tt"
		nameserver    = "127.0.0.63"
	)
	dnsOsConfigurator, err := NewOSConfigurator(func(format string, args ...interface{}) {
		log.Printf(format, args...)
	}, interfaceName)
	if err != nil {
		t.Fatalf("create dns os configurator: %v", err)
	}
	defer dnsOsConfigurator.Close()

	if dnsOsConfigurator.SupportsSplitDNS() {
		log.Println("SupportsSplitDNS")
	}

	fqdn, err := dnsname.ToFQDN(domainZone)
	if err != nil {
		panic(err)
	}
	newOSConfig := OSConfig{
		Nameservers:  []netaddr.IP{netaddr.MustParseIP(nameserver)},
		MatchDomains: []dnsname.FQDN{fqdn},
	}

	err = dnsOsConfigurator.SetDNS(newOSConfig)
	if err != nil {
		t.Errorf("set dns config to os configurator: %v", err)
	}
	fmt.Println("setup finished")

	time.Sleep(1 * time.Minute)
}
```

</details>

Run this test under root

Output:
```
=== RUN   TestNewManager
2021/10/19 23:29:45 dns: [rc=resolved nm=yes nm-resolved=yes nm-safe=no ret=systemd-resolved]
2021/10/19 23:29:45 SupportsSplitDNS
    test_test.go:42: set dns config to os configurator: setLinkDefaultRoute: Unknown method 'SetLinkDefaultRoute' or interface 'org.freedesktop.resolve1.Manager'.
setup finished
--- FAIL: TestNewManager (5.01s)
```

### Are there any recent changes that introduced the issue?

_No response_

### OS

Linux

### OS version

Ubuntu 18.04.6

### Tailscale version

master eaa0aef9344a86a1d06297dd2e1be6abe367d982

### Bug report

_No response_

--- a/net/dns/resolved.go
+++ b/net/dns/resolved.go
@@ -176,7 +176,13 @@ func (m *resolvedManager) SetDNS(config OSConfig) error {
 	}
 
 	if call := m.resolved.CallWithContext(ctx, "org.freedesktop.resolve1.Manager.SetLinkDefaultRoute", 0, m.ifidx, len(config.MatchDomains) == 0); call.Err != nil {
-		return fmt.Errorf("setLinkDefaultRoute: %w", call.Err)
+		if dbusErr, ok := call.Err.(dbus.Error); ok && dbusErr.Name == dbus.ErrMsgUnknownMethod.Name {
+			// on some older systems like Kubuntu 18.04.6 with systemd 237 method SetLinkDefaultRoute is absent,
+			// but otherwise it's working good
+			m.logf("[v1] failed to set SetLinkDefaultRoute: %v", call.Err)
+		} else {
+			return fmt.Errorf("setLinkDefaultRoute: %w", call.Err)
+		}
 	}
 
 	// Some best-effort setting of things, but resolved should do the

